AWSIoT/AWSIoTDataManager.m

```diff
@@ -342,7 +344,47 @@ - (void)addUserMetaData:(NSDictionary<NSString *, NSString *> *)userMetaDataMap
     if (userMetaDataMap) {
         for (id key in userMetaDataMap) {
             if (!([key isEqualToString:@"SDK"] || [key isEqualToString:@"Version"])) {
-                [userMetaDataString appendFormat:@"&%@=%@", key, [userMetaDataMap objectForKey:key]];
+                NSString *metaDataValue = [userMetaDataMap objectForKey:key];
+                if ([metaDataValue isEqualToString:@""] || metaDataValue == nil){
+                    [userMetaDataString appendFormat:@"&%@", key];
+                } else {
+                    [userMetaDataString appendFormat:@"&%@=%@", key, metaDataValue];
+                }
+            } else {
+                AWSDDLogWarn(@"Keynames 'SDK' and 'Version' are reserved and will be skipped");
+            }
+        }
+    }
+
+    NSUInteger baseLength = [userMetaDataString length];
+    if (baseLength > 255) {
+        AWSDDLogWarn(@"Total number of characters in username fields cannot exceed (%lu)", (255 - baseLength));
+        self.mqttClient.userMetaData = [userMetaDataString substringToIndex:255];
+    } else {
+        self.mqttClient.userMetaData = [NSString stringWithString:userMetaDataString];
+    }
+}
+
+- (void)updateUserMetaData:(NSDictionary<NSString *, NSString *> *)userMetaDataMap {
+
+    // update the user metadata dict
+    for (id key in userMetaDataMap) {
+        [_userMetaDataDict setObject:[userMetaDataMap objectForKey:key] forKey: key];
+    }
+
+    // validate the length of username field
+    NSMutableString *userMetaDataString = [NSMutableString stringWithFormat:@"%@%@", @"?SDK=iOS&Version=", AWSIoTSDKVersion];
+
+    // Append each of the user-specified key-value pair to the connection username
+    if (_userMetaDataDict) {
+        for (id key in _userMetaDataDict) {
```

Comment:
Can we combine the logic of constructing the `userMetaDataString`?

Reply:
Discussed offline